readme.md
# async-exit-hook
[![Build Status](https://api.travis-ci.org/Tapppi/async-exit-hook.svg)](https://travis-ci.org/Tapppi/async-exit-hook)
[![Coverage Status](https://coveralls.io/repos/github/Tapppi/async-exit-hook/badge.svg?branch=master)](https://coveralls.io/github/Tapppi/async-exit-hook?branch=master)

> Run some code when the process exits

The `process.on('exit')` event doesn't catch all the ways a process can exit. This module catches:

* process SIGINT, SIGTERM and SIGHUP, SIGBREAK signals  
* process beforeExit and exit events  
* PM2 clustering process shutdown message ([PM2 graceful reload](http://pm2.keymetrics.io/docs/usage/cluster-mode/#graceful-reload))  

Useful for cleaning up. You can also include async handlers, and add custom events to hook and exit on.

Forked and pretty much rewritten from [exit-hook](https://npmjs.com/package/exit-hook).


## Install

```
$ npm install --save async-exit-hook
```

## Usage

### Considerations and warning
#### On `process.exit()` and asynchronous code
**If you use asynchronous exit hooks, DO NOT use `process.exit()` to exit.
The `exit` event DOES NOT support asynchronous code.**
>['beforeExit' is not emitted for conditions causing explicit termination, such as process.exit()]
(https://nodejs.org/api/process.html#process_event_beforeexit)

#### Windows and `process.kill(signal)`
On windows `process.kill(signal)` immediately kills the process, and does not fire signal events, 
and as such, cannot be used to gracefully exit. See *Clustering and child processes* for a
workaround when killing child processes. I'm planning to support gracefully exiting 
with async support on windows soon.

### Clustering and child processes
If you use custom clustering / child processes, you can gracefully shutdown your child process
by sending a shutdown message (`childProc.send('shutdown')`).

### Example
```js
const exitHook = require('async-exit-hook');

exitHook(() => {
    console.log('exiting');
});

// you can add multiple hooks, even across files
exitHook(() => {
    console.log('exiting 2');
});

// you can add async hooks by accepting a callback
exitHook(callback => {
    setTimeout(() => {
        console.log('exiting 3');
        callback();
    }, 1000);
});

// You can hook uncaught errors with uncaughtExceptionHandler(), consequently adding 
// async support to uncaught errors (normally uncaught errors result in a synchronous exit).
exitHook.uncaughtExceptionHandler(err => {
    console.error(err);
});

// You can add multiple uncaught error handlers
// Add the second parameter (callback) to indicate async hooks
exitHook.uncaughtExceptionHandler((err, callback) => {
    sendErrorToCloudOrWhatever(err, success => {
        if (err) {
            console.error('Error sending to cloud: ', err.stack);
        } else {
            console.log('Sent err to cloud');
        }
        callback();
    });
});

// Add exit hooks for a signal or custom message:

// Custom signal
// Arguments are `signal, exitCode` (SIGBREAK is already handled, this is an example)
exitHook.hookEvent('SIGBREAK', 21);

// process event: `message` with a filter
// filter gets all arguments passed to *handler*: `process.on(message, *handler*)`
// Exits on process event `message` with msg `customShutdownMessage` only
exitHook.hookEvent('message', 0, msg => msg !== 'customShutdownMessage');

// All async hooks will work with uncaught errors when you have specified an uncaughtExceptionHandler
throw new Error('awesome');

//=> // Sync uncaughtExcpetion hooks called and retun
//=> '[Error: awesome]'
//=> // Sync hooks called and retun
//=> 'exiting'
//=> 'exiting 2'
//=> // Async uncaughtException hooks return
//=> 'Sent error to cloud'
//=> // Sync uncaughtException hooks return
//=> 'exiting 3'
```


## License

MIT © Tapani Moilanen  
MIT © [Sindre Sorhus](http://sindresorhus.com)


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="2.0.0"></a>
# [2.0.0](https://github.com/tapppi/async-exit-hook/compare/v1.1.2...v2.0.0) (2017-08-03)


### Features

* add unhandledRejectionHandler ([#3](https://github.com/tapppi/async-exit-hook/issues/3)) ([96a194f](https://github.com/tapppi/async-exit-hook/commit/96a194f))


### BREAKING CHANGES

* unhandledExceptionHandler no longer
catches rejections.



<a name="1.1.2"></a>
## [1.1.2](https://github.com/tapppi/async-exit-hook/compare/v1.1.1...v1.1.2) (2017-03-29)


### Bug Fixes

* filters are used individually for events [#1](https://github.com/tapppi/async-exit-hook/issues/1) ([03235c8](https://github.com/tapppi/async-exit-hook/commit/03235c8))



<a name="1.1.1"></a>
## [1.1.1](https://github.com/tapppi/async-exit-hook/compare/v1.1.0...v1.1.1) (2016-11-04)


### Bug Fixes

* unhandled rejections now handled ([4302b9e](https://github.com/tapppi/async-exit-hook/commit/4302b9e))


### Chores

* drop support for node 0.12 ([2830391](https://github.com/tapppi/async-exit-hook/commit/2830391))


### BREAKING CHANGES

* node 0.12 not tested anymore



<a name="1.1.0"></a>
# [1.1.0](https://github.com/tapppi/async-exit-hook/compare/v1.0.0...v1.1.0) (2016-10-13)


### Features

* support uncaughtRejectionHandler ([9098e3c](https://github.com/tapppi/async-exit-hook/commit/9098e3c))


test/tests.js
// Tests have to happen in a subprocess to test the exit functionality
'use strict';

const fork = require('child_process').fork;
const path = require('path');

const test = require('ava');

/**
 * Starts a test file in a subprocess, returns a promise that resolves with the subprocess
 * exit code and output in an array ([code, output])
 *
 * @async
 * @param {String} test Filename without path or extension
 * @param {String} signal Signal (or 'shutdown' for message) to send to the process
 * @return {Promise.<[Number, String]>} Array with the exit code and output of the subprocess
 */
function testInSub(test, signal) {
	return new Promise(resolve => {
		const proc = fork(
			path.resolve(__dirname, './cases/' + test + '.js'),
			{
				env: process.env,
				silent: true
			}
		);

		let output = '';

		proc.stdout.on('data', data => {
			output += data.toString();
		});

		proc.stderr.on('data', data => {
			output += data.toString();
		});

		proc.on('exit', code => {
			resolve([code, output]);
		});

		if (signal === 'shutdown') {
			proc.send(signal);
		} else if (signal) {
			proc.kill(signal);
		}
	});
}

test('API: test adding and removing and listing hooks', t => {
	const exitHook = require('./../');

	t.plan(3);

	// Enable hooks
	exitHook(() => {});

	// Ensure SIGBREAK hook
	exitHook.hookEvent('SIGBREAK', 128 + 21);
	t.not(exitHook.hookedEvents().indexOf('SIGBREAK'), -1);

	// Unhook SIGBREAK
	exitHook.unhookEvent('SIGBREAK');
	t.is(exitHook.hookedEvents().indexOf('SIGBREAK'), -1);

	// Rehook SIGBREAK
	exitHook.hookEvent('SIGBREAK', 128 + 21);
	t.not(exitHook.hookedEvents().indexOf('SIGBREAK'), -1);
});

test('sync handlers', async t => {
	t.plan(2);
	const [code, output] = await testInSub('sync', 'shutdown');

	t.is(output, 'SUCCESS');
	t.is(code, 0);
});

test('async handlers', async t => {
	t.plan(2);
	const [code, output] = await testInSub('async', 'shutdown');

	t.is(output, 'SUCCESS');
	t.is(code, 0);
});

test('async uncaught exception handler', async t => {
	t.plan(2);
	const [code, output] = await testInSub('async-err');

	t.is(output, 'SUCCESS');
	t.is(code, 0);
});

test('async exit timeout', async t => {
	t.plan(2);
	const [code, output] = await testInSub('async-exit-timeout');

	t.is(output, 'SUCCESS');
	t.is(code, 0);
});

test('unhandled promise rejection', async t => {
	t.plan(2);
	const [code, output] = await testInSub('unhandled-promise');

	t.is(output, 'SUCCESS');
	t.is(code, 0);
});


test/cases/stub.js
// Stub to make sure that the required callbacks are called by exit-hook
'use strict';
let c = 0;
let noCallback = true;

// Increment the called count
exports.called = () => {
	c++;
};

// Exit with error
exports.reject = (s, code) => {
	process.stdout.write('FAILURE: ' + s);
	// eslint-disable-next-line unicorn/no-process-exit
	process.exit(code === null || code === undefined ? 1 : code);
};

// Exit with success
exports.done = () => {
	process.stdout.write('SUCCESS');
	// eslint-disable-next-line unicorn/no-process-exit
	process.exit(0);
};

// Add the exit check with a specific expected called count
exports.addCheck = num => {
	noCallback = false;

	// Only call exit once, and save uncaught errors
	let called = false;
	let ucErrStr;

	// Save errors that do not start with 'test'
	process.on('uncaughtException', err => {
		if (err.message.indexOf('test') !== 0) {
			ucErrStr = err.stack;
		}
	});
	// Save rejections that do not start with 'test'
	process.on('unhandledRejection', reason => {
		if ((reason.message || reason).indexOf('test') !== 0) {
			ucErrStr = reason.message || reason;
		}
	});

	// Check that there were no unexpected errors and all callbacks were called
	function onExitCheck(timeout) {
		if (called) {
			return;
		}
		called = true;

		if (timeout) {
			exports.reject('Test timed out');
		} else if (ucErrStr) {
			exports.reject(ucErrStr);
		} else if (c === num) {
			exports.done();
		} else {
			exports.reject('Expected ' + num + ' callback calls, but ' + c + ' received');
		}
	}

	process.once('exit', onExitCheck.bind(null, null));
	setTimeout(onExitCheck.bind(null, true), 10000);
};

// If the check isn't added, throw on exit
process.once('exit', () => {
	if (noCallback) {
		exports.reject('FAILURE, CHECK NOT ADDED');
	}
});


test/cases/sync.js
'use strict';
const stub = require('./stub');
const exitHook = require('./../../index');

exitHook(() => {
	stub.called();
});

exitHook(() => {
	stub.called();
});

process.on('exit', () => {
	stub.called();
});

stub.addCheck(3);


index.js
'use strict';

var hooks = [];
var errHooks = [];
var called = false;
var waitingFor = 0;
var asyncTimeoutMs = 10000;

var events = [];
var filters = [];

function exit(exit, code, err) {
	// Only execute hooks once
	if (called) {
		return;
	}

	called = true;

	// Run hooks
	if (err) {
		// Uncaught exception, run error hooks
		errHooks.map(runHook.bind(null, 1, err));
	}
	hooks.map(runHook.bind(null, 0, null));

	if (waitingFor) {
		// Force exit after x ms (10000 by default), even if async hooks in progress
		setTimeout(function () {
			doExit();
		}, asyncTimeoutMs);
	} else {
		// No asynchronous hooks, exit immediately
		doExit();
	}

	// Runs a single hook
	function runHook(syncArgCount, err, hook) {
		// cannot perform async hooks in `exit` event
		if (exit && hook.length > syncArgCount) {
			// hook is async, expects a finish callback
			waitingFor++;

			if (err) {
				// Pass error, calling uncaught exception handlers
				return hook(err, stepTowardExit);
			}
			return hook(stepTowardExit);
		}

		// hook is synchronous
		if (err) {
			// Pass error, calling uncaught exception handlers
			return hook(err);
		}
		return hook();
	}

	// Async hook callback, decrements waiting counter
	function stepTowardExit() {
		process.nextTick(function () {
			if (--waitingFor === 0) {
				doExit();
			}
		});
	}

	var doExitDone = false;

	function doExit() {
		if (doExitDone) {
			return;
		}
		doExitDone = true;

		if (exit === true) {
			// All handlers should be called even if the exit-hook handler was registered first
			process.nextTick(process.exit.bind(null, code));
		}
	}
}

// Add a hook
function add(hook) {
	hooks.push(hook);

	if (hooks.length === 1) {
		add.hookEvent('exit');
		add.hookEvent('beforeExit', 0);
		add.hookEvent('SIGHUP', 128 + 1);
		add.hookEvent('SIGINT', 128 + 2);
		add.hookEvent('SIGTERM', 128 + 15);
		add.hookEvent('SIGBREAK', 128 + 21);

		// PM2 Cluster shutdown message. Caught to support async handlers with pm2, needed because
		// explicitly calling process.exit() doesn't trigger the beforeExit event, and the exit
		// event cannot support async handlers, since the event loop is never called after it.
		add.hookEvent('message', 0, function (msg) {
			if (msg !== 'shutdown') {
				return true;
			}
		});
	}
}

// New signal / event to hook
add.hookEvent = function (event, code, filter) {
	events[event] = function () {
		for (var i = 0; i < filters.length; i++) {
			if (filters[i].apply(this, arguments)) {
				return;
			}
		}
		exit(code !== undefined && code !== null, code);
	};

	if (!filters[event]) {
		filters[event] = [];
	}

	if (filter) {
		filters[event].push(filter);
	}
	process.on(event, events[event]);
};

// Unhook signal / event
add.unhookEvent = function (event) {
	process.removeListener(event, events[event]);
	delete events[event];
	delete filters[event];
};

// List hooked events
add.hookedEvents = function () {
	var ret = [];
	for (var name in events) {
		if ({}.hasOwnProperty.call(events, name)) {
			ret.push(name);
		}
	}
	return ret;
};

// Add an uncaught exception handler
add.uncaughtExceptionHandler = function (hook) {
	errHooks.push(hook);

	if (errHooks.length === 1) {
		process.once('uncaughtException', exit.bind(null, true, 1));
		process.once('unhandledRejection', exit.bind(null, true, 1));
	}
};

// Configure async force exit timeout
add.forceExitTimeout = function (ms) {
	asyncTimeoutMs = ms;
};

// Export
module.exports = add;
